setupext.py

```diff
@@ -960,43 +960,51 @@ def do_custom_build(self):
         else:
             libfreetype = 'libfreetype.a'
 
-        if os.path.isfile(os.path.join(src_path, 'objs', '.libs', libfreetype)):
+        # bailing because it is already built
+        if os.path.isfile(os.path.join(
+                src_path, 'objs', '.libs', libfreetype)):
             return
-        if not os.path.exists('build'):
-            os.makedirs('build')
-
-        url_fmts = [
-            ('https://downloads.sourceforge.net/project/freetype'
-             '/freetype2/{version}/{tarball}'),
-            ('https://download.savannah.gnu.org/releases/freetype'
-             '/{tarball}')
-        ]
-        tarball = 'freetype-{0}.tar.gz'.format(LOCAL_FREETYPE_VERSION)
-
-        target_urls = [
-            url_fmt.format(version=LOCAL_FREETYPE_VERSION,
-                           tarball=tarball)
-            for url_fmt in url_fmts]
-
-        for tarball_url in target_urls:
-            try:
-                tar_contents = download_or_cache(tarball_url,
-                                                 LOCAL_FREETYPE_HASH)
-                break
-            except Exception:
-                pass
-        else:
-            raise IOError("Failed to download FreeType.  Please download " +
-                          "one of {target_urls} ".format(
-                              target_urls=target_urls) +
-                          "and extract it into the build directory "
-                          "at the top-level of the source repository")
 
-        print("Building {}".format(tarball))
-        tar_contents.seek(0)
-        with tarfile.open(tarball, mode="r:gz", fileobj=tar_contents) as tgz:
-            tgz.extractall("build")
+        # do we need to download / load the source from cache?
+        if not os.path.exists(src_path):
+            if not os.path.exists('build'):
```

Comment:
this belongs out of the main if?  (if the source tree exists but not the buildddir, you still want to mkdir build)

Reply:
The src path is in `build` (see L954) so if we have the source dir, `build` exists.